tests/chainerx_tests/unit_tests/test_backward.py

```diff
@@ -555,3 +562,96 @@ def fprop(x0, x1):
     _check_grad(
         fprop, xs, tuple(expected_gxs), xs_indices=xs_indices,
         ys_indices=ys_indices)
+
+
+def test_grad_with_grad_outputs():
+    shape = (1,)
+    dtype = chainerx.float32
+
+    xs = (
+        chainerx.full(shape, 3, dtype).require_grad(),
+        chainerx.full(shape, 5, dtype).require_grad(),)
+
+    gysi = (
+        chainerx.full(shape, 10, dtype).require_grad(),
+        chainerx.full(shape, 20, dtype).require_grad(),)
+
+    expected_gxs = (
+        chainerx.full(shape, 110, dtype),
+        chainerx.full(shape, 70, dtype),)
+
+    def fprop(x0, x1):
+        return x0 + x1, x0 * x1
+
+    _check_grad(fprop, xs, expected_gxs, grad_outputs=gysi)
+
+
+def test_grad_with_invalid_grad_outputs():
+    shape = (1,)
+    dtype = chainerx.float32
+
+    xs = (
+        chainerx.full(shape, 3, dtype).require_grad(),
+        chainerx.full(shape, 5, dtype).require_grad(),)
+
+    gysi = (
+        chainerx.full(shape, 10, dtype).require_grad(),)
+
+    expected_gxs = (
+        chainerx.full(shape, 110, dtype),
+        chainerx.full(shape, 70, dtype),)
+
+    def fprop(x0, x1):
+        return x0 + x1, x0 * x1
+
+    with pytest.raises(chainerx.GradientError):
+        _check_grad(fprop, xs, expected_gxs, grad_outputs=gysi)
+
+
+def test_grad_with_set_grad():
+    shape = (1,)
+    dtype = chainerx.float32
+
+    xs = (
+        chainerx.full(shape, 3, dtype).require_grad(),
+        chainerx.full(shape, 5, dtype).require_grad(),)
+
+    expected_gxs = (
+        chainerx.full(shape, 6, dtype),
+        chainerx.full(shape, 4, dtype),)
+
+    def fprop(x0, x1):
+        return x0 + x1, x0 * x1
+
+    _check_grad(fprop, xs, expected_gxs, set_grad=True)
+
+
+def test_grad_with_retain_grad():
+    shape = (1,)
+    backprop_id = None
+    dtype = chainerx.float32
+
+    xs = (
+        chainerx.full(shape, 3, dtype).require_grad(),
+        chainerx.full(shape, 5, dtype).require_grad(),)
+
+    expected_gxs = (
+        chainerx.full(shape, 4, dtype),)
+
+    # If variables go out of scope
+    # then intermediate grads are wiped.
+    # We need to do all the computations here
+    a = xs[0]*2
+    b = a+xs[1]
+    c = a+b
```

Comment:
A nitpick but let's put spaces around the operators, for readability and consistency.